make it work with auxiliaryProjectTreeInfos kind of project + use the…

drzraf opened the pull request:
make it work with auxiliaryProjectTreeInfos kind of project + use the *real* root

kvch commented on wfw/lib.py:
```diff
@@ -20,11 +22,16 @@
     with open(TREE_DATA, 'r') as tree_data:
-        build(tree_data, ROOT)
+        mjson = json.load(tree_data)
+        if aux:
```
This is always going to be true, as `aux` is defined outside the function as true. I think`aux` should be passed to this function from `add_item`.

kvch commented:
Thanks for the PR! I checked your code and it does not seem to work at my machine, because of handling of `aux`. Could you please fix it? :)

drzraf commented:
that's right.

But allowing the user to define means modifying almost every function in
the lib.py because the root is used very deep into the call-stack and
build_tree_from_file() is called by most of them.

In my case, the Workflowy is "shared", maybe this could be detected, but
I feel saner to make it user-defined for now.
(maybe something to grab from ~/.wfwrc ?)


Don't you want to make tree.py a class, and initially define an instance
where "aux" would be a class property?
(build_tree_from_file could be a method of that "Tree" class)

kvch commented:
It is a good idea to create a tree class instead on the function "collection" in tree.py. :) The functions in that file could be the functions of the tree class.

Just to clear things up, is this `aux` handling is required when the Workflowy is shared with someone else? In this case, I agree with you. I would not expect anyone who has a shared tree to type `--aux` every time he/she uses `wfw`.

drzraf commented:
I think that shared-workflowy uses this slightly distinct tree structure
although I can't tell you for sure why/when/how.

I could provide you with a sample top-level tree, by email if you really
want to but from what I saw only this 2 level json-path changed.

I bet there is a way to use a variable-depth list index in python, but
can't tell you how.
I would have preferred to patch with something like:
> x = list["root"]<variable-depth-index>["leaf"]

Please note (cf other issue), that pushing change to workflowy.com does
not work (yet?)

kvch commented:
Thanks, but no need to send it to me. I validated the theory with a test list. :)

Why do you want to use a variable-depth list index exactly?

drzraf commented:
the variable-index could have been passed as a parameter as a "where to
search in the tree" rather than this higher-level concept
"is-workflowy-auxiliary" put deeply into the structure internals.

But it's debatable and not that important.
If it just works, it'll be good enough.

kvch commented:
@drzraf Are you planning to address the issues discussed above?

drzraf commented:
AFAICT
The only real issue is handling "aux".
1. Modifying every function prototype is intrusive and not good in the long-run.
1. Having a class is the way to go.

But 2. means that it is somehow a "re-factoring"... for which I'm sure you would feel more confident doing it yourself.

Moreover, as long as I'm affected by #4 I'm not that keen of making big code changes for a tool that I can't even fully test on my own WF.

kvch commented:
Okay, I will look into #4 and do the refactor. Thanks. :)

kvch commented:
Could you rebase the branch?

kvch commented:
@drzraf I did the refactor. I think only your aux data fix is needed.

drzraf commented:
I couldn't find the (public?) branch with the refactor.

kvch commented:
It's on master.

Commit:
fix a regression from refactor during 517993a7

Commit:
support shared tree: 2/3: pass the "shared" ~/wfwrc value to the Tree builder so that the correct root can be grabbed

Commit:
support shared tree: 3/3: changed the way root is passed along to post_local_change() so that the correct root is used during API calls

Commit:
support shared tree: 1/3: added a "shared" parameter to Tree constructor
* New method: find_data_root() is a helper which return the root according
  to the expected Tree structure (mainProjectTreeInfo vs auxiliaryProjectTreeInfos)
* shared and tree_data now (read-only) properties of the Tree object

Commit:
Tree: stop using "My list" as a hardcoded root name/id

Commit:
error handling when adding to an unknown parent

kvch commented on wfw/lib.py:
```diff
@@ -26,7 +26,7 @@ def get_user_data():
         user_data['email'] = configparser.get('user', 'email')
         user_data['password'] = configparser.get('user', 'password')
-        user_data['aux'] = configparser.get('tree', 'shared')
+        user_data['shared'] = configparser.get('tree', 'shared')
```
It should be `configparser.getboolean`, because now the value is string, so checking if `self.shared` is `True` always evaluates to true. My bad.

kvch commented on wfw/tree.py:
```diff
@@ -13,13 +13,35 @@
+    def __init__(self, tree_data, shared):
+        self._shared = shared
+        self._tree_data = tree_data
+        root = self.find_data_root()['rootProject']
```
In case of my tree root is `None`, that is why I had hardcoded values for the root.

```{u'dateJoinedTimestampInSeconds': 00000000,
 u'initialMostRecentOperationTransactionId': u'00000000',
 u'initialPollingIntervalInMs': 10000,
 u'isReadOnly': False,
 u'itemsCreatedInCurrentMonth': 0,
 u'monthlyItemQuota': 0,
 u'ownerId': 000000,
 u'rootProject': None,
 u'rootProjectChildren': [{u'ch': [{u'id': u'21dsfs537-cdsdf-da74-sdfd9-09cxcvbthaea3',
                                    u'lm': 000000000,
                                    u'nm': u'text'
                                   {u'id': u'apsdfisad-004b-4a6c-8129-sdfasdfasdf',
```

kvch reviewed (changes requested):
In case of my tree, the `rootProject`
